### ISSUE TITLE
roslaunch take much time before starting nodes

### ISSUE THREAD

**Issue by garaemon** [ros/ros_comm#535]
roslaunch takes much time before starting nodes.
environment: SSD, hydro, ubuntu 12.04

I run roslaunch with cProfile as `python -m cProfile roslaunch` and here is the result of profiling with
`roslaunch openni2_launch openni2.launch`
https://gist.github.com/garaemon/82127eade788f9778f50

According to the profiling, realpath took 17 seconds...

**Comment by garaemon** [comment]
It might be a problem of catkin_pkg

**Comment by dirk-thomas** [comment]
The benchmark results will vary significantly based on the environment. Please post more information about your test scenario, all the environment variable, etc. It is very likely that by changing your ROS_PACKAGE_PATH and/or ignoring certain highly nested subfolders  the performance can be improved significantly in your specific case.

I also don't think that this will be resolvable in general. Simply because currently any lookup of ROS resources requires crawling the file system. The design can't be changes easily and will require a fundamentally different approach (as it has been consider for ROS 2: https://github.com/ament/ament_cmake/blob/master/ament_cmake_core/doc/resource_index.md).

**Comment by garaemon** [comment]
I found that I have a lot of packages under workspace directory and If I use /opt/ros/hydro/setup.bach, roslaunch is fast.

```
$ rospack list | wc -l
750
$ source /opt/ros/hydro/setup.zsh; rospack list | wc -l
256
```

**Comment by garaemon** [comment]
I found [this](https://github.com/ros-infrastructure/catkin_pkg/pull/111) eases the issue.

```
$ echo $ROS_PACKAGE_PATH
/home/garaemon/ros/hydro/src:/opt/ros/hydro/share:/opt/ros/hydro/stacks
$ find ~/ros/hydro/src | wc -l
53680
$ find ~/ros/hydro/src -name package.xml | wc -l
221
$ find ~/ros/hydro/src -name manifest.xml | wc -l
188
```

**Comment by dirk-thomas** [comment]
Since roslaunch is fast when only using `/opt/ros/hydro` is seems to be related to your workspace. Can you post e.g. a rosinstall file with the repositories you have checked out in your workspace?

**Comment by garaemon** [comment]
I uploaded .rosinstall [here](https://gist.github.com/garaemon/66723c8ce19dea680fc9)

**Comment by dirk-thomas** [comment]
The URLs of the svn entries of the posted rosinstall file are invalid.

Anyway the generated workspace seems to be much smaller for me:

```
find . | wc -l
16868
```

These seems to be files in your workspace not covered by the posted rosinstall file.

**Comment by garaemon** [comment]
I have non-public one repository...

**Comment by dirk-thomas** [comment]
Without seeing the workspace I can only give two generic advises:
- the workspace might contain a deeply nested folder structure with a package.xml / manifest.xml
- the launch file might contain a specific entry which result in crawling a deeply nested file system inside a single package

If you could dig into this and try to figure out the exact reason it might be possible to consider a workaround. Otherwise there is not much I can do.

**Comment by garaemon** [comment]
I agree that nested filesystem causes this problem. 
Deeply nested filesystem requires calling of `os.path.realpath` many times and it takes much time.

**Comment by garaemon** [comment]
I will dig my filesystem structure to find "cancer" of my environment.

**Comment by garaemon** [comment]
I will open new issue
